pkg/azure/validation/validation_test.go

```diff
@@ -0,0 +1,477 @@
+			ConsistOf(
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("data.clientID")})),
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("data.clientSecret")})),
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("data.subscriptionID")})),
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("data.tenantID")})),
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("data.userData")})),
+			),
+		},
+		{"should succeed when all required fields are present", testClientID, testClientSecret, testSubscriptionID, testTenantID, testUserData, 0, nil},
+	}
+
+	g := NewWithT(t)
+	for _, entry := range table {
+		t.Run(entry.description, func(t *testing.T) {
+			secret := createSecret(entry.clientID, entry.clientSecret, entry.subscriptionID, entry.tenantID, entry.testUserData)
+			errList := ValidateProviderSecret(secret)
+			g.Expect(len(errList)).To(Equal(entry.expectedErrors))
+			if entry.matcher != nil {
+				g.Expect(errList).To(entry.matcher)
+			}
+		})
+	}
+}
+
+func TestValidateSubnetInfo(t *testing.T) {
+	const (
+		testSubnetName = "test-control-ns-nodes"
+		testVnetName   = "test-control-ns"
+	)
+
+	fldPath := field.NewPath("providerSpec", "subnetInfo")
+
+	table := []struct {
+		description    string
+		vnetName       string
+		subnetName     string
+		expectedErrors int
+		matcher        gomegatypes.GomegaMatcher
+	}{
+		{"should forbid empty vnetName",
+			"", testSubnetName, 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.subnetInfo.vnetName")}))),
+		},
+		{"should forbid empty subnetName",
+			testVnetName, "", 1,
+			ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.subnetInfo.subnetName")}))),
+		},
+		{"should forbid empty subnetName and vnetName",
+			"", "", 2,
+			ConsistOf(
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.subnetInfo.vnetName")})),
+				PointTo(MatchFields(IgnoreExtras, Fields{"Type": Equal(field.ErrorTypeRequired), "Field": Equal("providerSpec.subnetInfo.subnetName")})),
+			),
+		},
+		{"should succeed when vnetName and subnetName are present", testVnetName, testSubnetName, 0, nil},
+	}
+	g := NewWithT(t)
+	for _, entry := range table {
+		t.Run(entry.description, func(t *testing.T) {
+			subnetInfo := api.AzureSubnetInfo{
+				VnetName:   entry.vnetName,
+				SubnetName: entry.subnetName,
+			}
+			errList := validateSubnetInfo(subnetInfo, fldPath)
+			g.Expect(len(errList)).To(Equal(entry.expectedErrors))
+			if entry.matcher != nil {
+				g.Expect(errList).To(entry.matcher)
+			}
+		})
+	}
+
+}
+
+func TestValidateHardwareProfile(t *testing.T) {
```

Comment:
```suggestion
func TestValidateNonEmptyVMSizeInHardwareProfile(t *testing.T) {
```

Reply:
HardwareProfile cannot be empty so having NonEmpty in the name does not add any additional value but confuses because now it means that it can be empty and there should potentially be a test for that too.